src/coreclr/jit/likelyclass.cpp

```diff
@@ -168,84 +171,98 @@ extern "C" DLLEXPORT CORINFO_CLASS_HANDLE WINAPI getLikelyClass(ICorJitInfo::Pgo
 
                 case 2:
                 {
+                    // Fast path for two classes
                     if ((h.HistogramEntryAt(0).m_count >= h.HistogramEntryAt(1).m_count) &&
                         !ICorJitInfo::IsUnknownTypeHandle(h.HistogramEntryAt(0).m_mt))
                     {
-                        *pLikelihood = (100 * h.HistogramEntryAt(0).m_count) / h.m_totalCount;
-                        return (CORINFO_CLASS_HANDLE)h.HistogramEntryAt(0).m_mt;
-                    }
-                    else if (!ICorJitInfo::IsUnknownTypeHandle(h.HistogramEntryAt(1).m_mt))
-                    {
-                        *pLikelihood = (100 * h.HistogramEntryAt(1).m_count) / h.m_totalCount;
-                        return (CORINFO_CLASS_HANDLE)h.HistogramEntryAt(1).m_mt;
+                        pLikelyClasses[0].likelihood = (100 * h.HistogramEntryAt(0).m_count) / h.m_totalCount;
+                        pLikelyClasses[0].clsHandle  = (CORINFO_CLASS_HANDLE)h.HistogramEntryAt(0).m_mt;
+
+                        if (!ICorJitInfo::IsUnknownTypeHandle(h.HistogramEntryAt(1).m_mt))
+                        {
+                            pLikelyClasses[1].likelihood = (100 * h.HistogramEntryAt(1).m_count) / h.m_totalCount;
+                            pLikelyClasses[1].clsHandle  = (CORINFO_CLASS_HANDLE)h.HistogramEntryAt(1).m_mt;
+                            return 2;
+                        }
+                        return 1;
                     }
-                    else
+
+                    if (!ICorJitInfo::IsUnknownTypeHandle(h.HistogramEntryAt(1).m_mt))
                     {
-                        return NULL;
+                        pLikelyClasses[0].likelihood = (100 * h.HistogramEntryAt(1).m_count) / h.m_totalCount;
+                        pLikelyClasses[0].clsHandle  = (CORINFO_CLASS_HANDLE)h.HistogramEntryAt(1).m_mt;
+
+                        if (!ICorJitInfo::IsUnknownTypeHandle(h.HistogramEntryAt(0).m_mt))
+                        {
+                            pLikelyClasses[1].likelihood = (100 * h.HistogramEntryAt(0).m_count) / h.m_totalCount;
+                            pLikelyClasses[1].clsHandle  = (CORINFO_CLASS_HANDLE)h.HistogramEntryAt(0).m_mt;
+                            return 2;
+                        }
+                        return 1;
                     }
+                    return 0;
                 }
-                break;
 
                 default:
                 {
-                    // Find maximum entry and return it
-                    //
-                    unsigned maxKnownIndex = 0;
-                    unsigned maxKnownCount = 0;
-
+                    // Since this method can be invoked without a jit instance we can't use any existing allocators
+                    auto sortedEntries = (LikelyClassHistogramEntry*)malloc(sizeof(LikelyClassHistogramEntry) *
+                                                                            h.countHistogramElements);
```

Comment:
Should we use stack memory when the count is low? I think the histogram size is currently at most 8, but it can be larger when merging different profiles, so maybe a threshold like 64 or something would ensure we almost never need to heap alloc.